produce.go

```diff
@@ -84,133 +100,35 @@ func (cmd *produceCmd) run(args []string) error {
-}
-
-func kafkaCompressionForType(codecName string) (sarama.CompressionCodec, error) {
-	switch codecName {
-	case "gzip":
-		return sarama.CompressionGZIP, nil
-	case "snappy":
-		return sarama.CompressionSnappy, nil
-	case "lz4":
-		return sarama.CompressionLZ4, nil
-	case "":
-		return sarama.CompressionNone, nil
-	}
-	return sarama.CompressionNone, fmt.Errorf("unsupported compression codec %#v - supported: gzip, snappy, lz4", codecName)
-}
-
-func (cmd *produceCmd) findLeaders() error {
-	cfg, err := cmd.saramaConfig("produce")
-	if err != nil {
-		return err
-	}
-	cfg.Producer.RequiredAcks = sarama.WaitForAll
-
-loop:
-	for _, addr := range cmd.brokers {
-		broker := sarama.NewBroker(addr)
-		if err = broker.Open(cfg); err != nil {
-			warningf("failed to open broker connection to %v: %v", addr, err)
-			continue loop
-		}
-		if connected, err := broker.Connected(); !connected || err != nil {
-			warningf("Failed to open broker connection to %v: %v", addr, err)
-			continue loop
-		}
-
-		res, err := broker.GetMetadata(&sarama.MetadataRequest{Topics: []string{cmd.topic}})
-		if err != nil {
-			warningf("Failed to get metadata from %#v: %v", addr, err)
-			continue loop
-		}
-
-		brokers := map[int32]*sarama.Broker{}
-		for _, b := range res.Brokers {
-			brokers[b.ID()] = b
-		}
-
-		for _, tm := range res.Topics {
-			if tm.Name == cmd.topic {
-				if tm.Err != sarama.ErrNoError {
-					warningf("Failed to get metadata from %#v: %v", addr, tm.Err)
-					continue loop
-				}
-
-				cmd.leaders = map[int32]*sarama.Broker{}
-				for _, pm := range tm.Partitions {
-					b, ok := brokers[pm.Leader]
-					if !ok {
-						return fmt.Errorf("failed to find leader in broker response, giving up")
-					}
-
-					if err = b.Open(cfg); err != nil && err != sarama.ErrAlreadyConnected {
-						return fmt.Errorf("cannot open broker connection: %v", err)
-					}
-					if connected, err := broker.Connected(); !connected && err != nil {
-						return fmt.Errorf("failed to wait for broker connection to open: %v", err)
-					}
-
-					cmd.leaders[pm.ID] = b
-				}
-				return nil
-			}
-		}
-	}
-	return fmt.Errorf("failed to find leader for given topic")
-}
-
-func (cmd *produceCmd) close() {
-	for _, b := range cmd.leaders {
-		connected, err := b.Connected()
-		if err != nil {
-			warningf("Failed to check if broker is connected: %v", err)
-			continue
-		}
-		if !connected {
-			continue
-		}
-		if err := b.Close(); err != nil {
-			warningf("Failed to close broker %v connection: %v", b, err)
-		}
-	}
+	return cmd.produce(messages)
 }
 
 func (cmd *produceCmd) deserializeLines(in chan string, out chan producerMessage, partitionCount int32) {
 	defer close(out)
-	partition := int32(cmd.partition)
 	for l := range in {
 		l := l
 		var msg producerMessage
 
 		if cmd.literal {
 			data, _ := json.Marshal(l) // Marshaling a string can't fail.
 			msg.Value = json.RawMessage(data)
-			msg.Partition = &partition
+			// TODO allow specifying a key?
```

Comment:
This should be turned into an issue, wdyt?

Reply:
I've already got a PR for it, so I'm not sure that's worth it.